Add TraceState to SpanContext

awiede opened the pull request:
Resolves #211 

Adds `TraceState` to `SpanContext` and gives `TraceState` new methods per [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracestate).

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -111,6 +114,110 @@ impl SpanId {
+#[cfg_attr(feature = "serialize", derive(Deserialize, Serialize))]
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct TraceState {
+    data: HashMap<String, String>,
```
If this is an ordered map, could it simply be a `BTreeMap`?

awiede replied:
The problem is that the sorting is by insertion order, not key. So we could just keep everything in a `Vec` but then you have to iterate through everything on `get`. Rust doesn't have something like [Boost multi index](https://www.boost.org/doc/libs/1_62_0/libs/multi_index/doc/index.html), which would be ideal. There's an upper bound on how long trace state can be in the spec, so maybe it's not a big deal to just make this a wrapper around a `VecDeque` and `get` can just be `O(n)` where `n` isn't that big.

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -111,6 +114,110 @@ impl SpanId {
+impl TraceState {
+    /// Creates a new `TraceState` from the given key-value collection.
+    pub fn from_key_value(trace_state: Vec<KeyValue>) -> Self {
+        let mut data: HashMap<String, String> = HashMap::with_capacity(trace_state.len());
```
If you were to go with `BTreeMap` this could be simplified to something like:

```rust
pub fn new<T: IntoIterator<Item = (String, String)>>(trace_state: T) -> Self {
    TraceState {
        data: trace_state.into_iter().collect(),
    }
}
```

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -111,6 +114,110 @@ impl SpanId {
+    }
+
+    /// Retrieves a value for a given key from the `TraceState` if it exists.
+    pub fn get(&self, key: String) -> Option<String> {
```
This could probably return `Option<&str>`, let the caller `clone` if they want owned data

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -111,6 +114,110 @@ impl SpanId {
+                "{}{}{}",
+                key,
+                entry_delimiter,
+                self.data.get(key.as_str()).cloned().unwrap()
```
could do `as_ref().unwrap()` here to avoid the clone.

If you go with `BTreeMap` this also could be simplified to:

```rust
for (key, value) in self.data.iter() {
    ordered.push(format!("{}{}{}", key, entry_delimiter, value))
}
```

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -178,6 +299,11 @@ impl SpanContext {
     }
+
+    /// Returns the context's `TraceState`.
+    pub fn trace_state(&self) -> TraceState {
```
This could also return `&TraceState` to let the caller clone.

jtescher reviewed (changes requested):
Great start @awiede! added a few ideas for potential allocation improvements

awiede commented:
@jtescher - I _think_ I addressed all of your comments. RE: using a `BTreeMap`, I don't think that will work since we aren't sorting on the keys, but instead on the insertion order. Given that this should be a relatively small data set, I updated to just use an ordered collection of tuples. `TraceState::get` is now `O(n)` instead `O(constant)`, _but_ given the limitations on the size of `tracestate` itself (I think the spec says 32 elements), I don't know if that's a big deal.

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -111,6 +113,116 @@ impl SpanId {
+    }
+
+    /// Retrieves a value for a given key from the `TraceState` if it exists.
+    pub fn get(&self, key: String) -> Option<&str> {
```
This could probably take an `&str` to avoid callers having to clone.

jtescher commented on src/api/trace/span_context.rs:
```diff
@@ -111,6 +113,116 @@ impl SpanId {
+    /// given key, this updates the value and updates the value's position. Returns the updated
+    /// `TraceState`.
+    pub fn insert(self, key: String, value: String) -> TraceState {
+        let mut trace_state = self.delete(key.clone());
```
I just re-read the spec and it looks like:

> All mutating operations MUST return a new TraceState with the modifications applied. 

So we'll have to clone the trace state instead of modifying in place looks like.

Also 

> If invalid value is passed the operation MUST NOT return TraceState containing invalid data

So looks like these methods have to return `Result` types and do some validating

awiede replied:
So `insert` and `delete` are `FnOnce` functions, I can have them return a `Result<TraceState, ()>`, but that would mean on error the original valid `TraceState` has been moved/lost. I could do `Result<TraceState, TraceState>` which would return a clone of the unmodified input which I guess is better than silently ignoring an operation with an invalid key/value?

jtescher replied:
I think they have to be `Fn` functions and take `&self` to be spec compliant, unless I'm misunderstanding the `All mutating operations MUST return a new TraceState with the modifications applied.` part.

jtescher replied:
Seems like trace state is meant to be immutable so you have copies with the changes instead of mutating the underlying values

jtescher commented on src/api/trace/trace_context_propagator.rs:
```diff
@@ -171,20 +152,21 @@ impl api::TextMapFormat for TraceContextPropagator {
 
     #[rustfmt::skip]
-    fn extract_data() -> Vec<(&'static str, api::SpanContext)> {
+    fn  extract_data() -> Vec<(&'static str, &'static str, api::SpanContext)> {
```
this first whitespace was probably not intended.

awiede replied:
Ooops, serves me right for skipping formatting on this function 😞

jtescher reviewed (changes requested):
@awiede yeah that makes sense I hadn't read the spec all that closely 😬, looked it over and noticed a few other changes we may want to do

awiede commented:
@jtescher updated to `Fn` for `insert` and `delete`. Spec technically says that keys should be lowercase which leads to some annoying conflicts with the AWS propagator. I think it's fine for now since the AWS reserved keys only capitalize the first character, custom camel cased keys might get hairy.

jtescher reviewed (approved):
Looks good thanks @awiede
